### PR TITLE
Refactor which resolver: OS-specific workspace fallback

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds tests and simplifies which resolver logic by extracting workspace fallback into its own OS-aware module and centralizing path handling.
- Introduces per-platform backends for workspace lookups, consolidates candidate resolution, and reduces per-platform branching in the direct path resolution path.
- Replaces the former workspace.rs with a new workspace/mod.rs wiring in platform-specific backends (posix.rs and windows.rs).
- Updates tests, fixtures, and docs to reflect the new workspace-aware structure and usage.

## Changes
- Core resolver refactor
  - Introduced a dedicated workspace fallback module under src/stdlib/which/lookup/workspace with OS-specific backends (posix.rs and windows.rs).
  - Replaced the former workspace.rs with a new workspace/mod.rs that wires in the platform-specific backends.
  - Added a platform-agnostic candidates_for_dir helper to centralize candidate resolution per directory.
  - Centralized direct path handling with normalize_direct_path and is_executable checks; removed per-platform branching in resolve_direct where feasible.
  - Added should_visit_entry and workspace fallback max depth/skip configuration constants to the new workspace module.
  - Retained and reused the EnvSnapshot-based testing approach, but updated calls to the new workspace API shape.

- Tests and test infrastructure
  - Expanded test support in src/stdlib/register.rs to cover more non-Unix file types (pipes, block/char devices) where supported.
  - Reworked tests under src/stdlib/which/lookup/tests.rs to align with the new workspace module and EnvSnapshot usage; added tests for path-with-invalid-UTF-8 and relative PATH entries resolving against CWD.
  - Introduced a new workspace-aware test module setup under tests/std_filter_tests/which_filter_common.rs and adjusted which tests to use a reusable WhichTestFixture, simplifying test fixtures and improving test isolation.
  - Updated integration tests in tests/std_filter_tests/which_filter_tests.rs to reflect the new fixture and workspace fallback behavior, including cases for fresh lookups and cached results when executables are removed.
  - Added tests for path resolution with relative entries and PATHEXT handling where relevant (Windows).

- Documentation
  - docs/netsuke-design.md updated to document workspace fallback traversals, max depth, heavy-directory skipping, and NETSUKE_WHICH_WORKSPACE env var to disable the fallback for latency-sensitive scenarios.

## Rationale
- Clear separation of concerns: moving workspace fallback logic into its own module with platform-specific backends clarifies responsibilities and makes the resolver easier to maintain.
- Performance and safety: the workspace fallback is bounded (depth 6) and skips heavy directories to prevent excessive traversal on large trees.
- Consistent testing surface: expanding tests for PATH handling, invalid UTF-8 in PATH, and relative PATH entries improves reliability across Unix and Windows platforms.
- Aligns tests with the refactored codepath: the new fixtures and test helpers reflect the updated resolver signatures and behavior.

## Compatibility / Breaking changes
- Internal API changes to the which/lookup workspace handling. The workspace fallback API now lives in a dedicated module with OS-specific backends. Tests and code references were updated accordingly.
- Some non-Unix and Windows-specific code paths have been restructured; behavior remains functionally similar, with additional coverage for edge cases.

## How to test
- Run the full test suite: cargo test
- Pay attention to Windows-specific tests around PATHEXT and workspace traversal; ensure PATH and PATHEXT are set in environments accordingly when running Windows-targeted tests.
- Validate edge cases added in tests:
  - PATH entries with invalid UTF-8 components
  - Relative PATH entries resolving against the current working directory
  - Direct path resolution when PATH is not used, including canonicalisation behavior when fresh lookups are requested
- Review updated docs to see the documented workspace fallback behavior and environment variable controls.

## Notes
- No user-facing API changes for end users; internal resolver internals and test scaffolding have been updated to improve maintainability and coverage.
- The changes lay groundwork for more exhaustive cross-platform tests around which resolution, caching, and workspace fallbacks in future iterations.

📎 Task: https://www.terragonlabs.com/task/093f7066-ea7d-4255-8c8f-e7f02453291d

## Summary by Sourcery

Refactor the which resolver to use an OS-aware workspace fallback module, centralize path candidate handling, and extend tests and docs around PATH, PATHEXT, and workspace traversal behavior.

Enhancements:
- Introduce a workspace::mod module with POSIX and Windows backends to encapsulate workspace fallback traversal, skipping heavy directories and allowing it to be disabled via an environment variable.
- Centralize directory candidate generation and direct-path normalization in the which lookup logic to reduce platform-specific branching and improve maintainability.
- Extend non-Unix file predicates to cover pipes and device types, returning sensible defaults on unsupported platforms.

Documentation:
- Document workspace fallback depth limits, directory skipping, and the NETSUKE_WHICH_WORKSPACE environment flag in netsuke-design.md.

Tests:
- Extract shared which filter test fixtures into a reusable helper module and update integration tests to validate cache behavior, JSON list outputs, cwd_mode handling, and duplicate path canonicalization.
- Broaden workspace lookup tests to cover invalid UTF-8 in PATH, relative PATH entries resolving against the current working directory, PATHEXT normalization on Windows, and direct-path executability errors on Unix.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/stdlib/which/lookup/workspace/windows.rs](https://github.com/leynos/netsuke/pull/241/files/787f979dd2151e1aa4dbaa553dee6a294774b1ce#diff-1875864c43453001505e390b3ccba02fb5dc04958d9d3fc5b57701b86bae94e8)

Comment on lines +128 to +142

~~~rust
fn should_stop_collecting(
    matches: &mut Vec<Utf8PathBuf>,
    entry: walkdir::DirEntry,
    command: &str,
    ctx: &WorkspaceMatchContext,
    collect_all: bool,
) -> Result<bool, Error> {
    if let Some(path) = process_workspace_entry(entry, command, ctx)? {
        matches.push(path);
        if !collect_all {
            return Ok(true);
        }
    }
    Ok(false)
}
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/69281/delta?repo-id=1160466&review-id=241&biomarker=Excess+Number+of+Function+Arguments&filename=src%2Fstdlib%2Fwhich%2Flookup%2Fworkspace%2Fwindows.rs&method=should_stop_collecting)
should_stop_collecting has 5 arguments, max arguments = 4

File: tests/std_filter_tests/which_filter_tests.rs [hunk 1]
```diff
@@ -1,190 +1,103 @@
...
-            path_guard,
-            #[cfg(windows)]
-            pathext_guard,
-        })
+    fixture.state.reset_impure();
+    let second_result = fixture.render(second_template);
+
+    if expect_second_err {
+        let err = second_result.expect("expected fresh which lookup to fail after removal");
+        assert!(err.to_string().contains("not_found"));
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/std_filter_tests/which_filter_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: src/stdlib/which/lookup/workspace/posix.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+use camino::Utf8PathBuf;
+use minijinja::{Error, ErrorKind};
+use walkdir::WalkDir;
+
+use super::super::is_executable;
+use super::{WORKSPACE_MAX_DEPTH, WORKSPACE_SKIP_DIRS, should_visit_entry};
+use crate::stdlib::which::env::EnvSnapshot;
+
+pub(super) fn search_workspace(
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/stdlib/which/lookup/workspace/posix.rs]: Has this now been resolved?
